ros/cc_defs.bzl

```diff
@@ -23,6 +23,10 @@ def cc_ros_library(name, ros_package_name = None, **kwargs):
     cc_library(
         name = name,
         local_defines = local_defines,
+        target_compatible_with = [
+            "@platforms//os:linux",
```

Comment:
Why do we need this? It's expected to work on Linux and OSX (maybe?) anyway?

Reply:
But it does not work on platforms such as qnx.

Reply:
The readme says it's tested on Linux (Ubuntu) and that's in principle good enough for me (what's tested is expected to work, what's not tested may work or not). I don't feel going through the repo and marking most of the targets working on Linux and OSX. Unless this is something of very high importance, I'd say we close this PR and maybe improve docs.

Reply:
The docs are very clear and I am aware that ROS in Linux only. It is also not highly important since there is always the option to patch things. However I am in favor of upstream contributions to share fixes and improvements with the community.

The reason for this PR is to express this "linux only" in code. Speaking so, we could even remove osx as platform to better match the docs. Without this PR, users would have to find other ways to not build ROS related targets on other platform. But usually you want to be able to do `bazel build //...` on every platform and the targets specify for which platforms they available. Or how would you address this problem?

Reply:
This approach is how you tell bazel in its own language where these libraries are executable and where not. It replaces the current, custom logic that errors out on the wrong platform, plus it allows bazel to smartly skip ros targets on other platforms that are not supported when building the `all` target.

Without this change, a monorepo that supports certain targets to be compiled on windows, will always fail on the ros libraries when using `bazel build ...`

So this is more of a improvement/replacement for [the current logic](https://github.com/mvukov/rules_ros/blob/9b00aefe7ba84e6e22b069be461a4948f95ca32f/ros/repositories/ros_comm.BUILD.bazel#L109).

Reply:
:+1: It's good that https://github.com/mvukov/rules_ros/blob/9b00aefe7ba84e6e22b069be461a4948f95ca32f/ros/repositories/ros_comm.BUILD.bazel#L109 is referenced here. So, the current state of the repo is that if you build for targets other than Linux and macos, the builds are going to fail (+ one gets nice msg `Only Linux and macOS are supported!`) -- since ros C++ nodes transitively depend on `roscpp_lib` that depends on `config_h`.

Do we need this PR then?

BTW, not sure about macos any more. I tried this on my old intel mac years ago and then it worked.

Reply:
I still think we need this PR.

You clearly specify that this is linux (and maybe osx) only and even provide a nice error message. That is great. However, this does not really solve any problem in a large codebase where these rules are used.

Let's say, we want to add a new platform which is windows. For this, you probably want to add a new CI job that does `bazel test //...` (or even your developers might want to do this locally). Then you will get the nice error message that the ros targets do not work on windows. You now have 2 options to solve this problem.

1. Adapt you CI job to `bazel test -- //... -//some/path/with/ros... -//another/with/ros/... //some/path/with/ros/but/here/is/some/windows/...`. This seems not like a very convenient option. In particular if someone adds new windows code under `//another/with/ros/...` and gets a green CI, simply because it is not built.
2. Use `target_compatible_with` and let the bazel targets specify where they can be used.

Comparing both approaches, option 2 seems like the better solution to me. What do you think?

Reply:
You're right, `bazel test //...` won't work as-is, nor `bazel cquery`. Thanks for pointing this out. For me it's sufficient to add target_compatible_with on roscpp_lib target as that it known to work only on Linux and macos. Not sure about `roscpp_traits`. I wouldn't put those constraints on cc_ros_{binary, library} macros -- the constraints should go IMO on targets that are problematic.

Reply:
According to https://www.ros.org/reps/rep-0003.html#noetic-ninjemys-may-2020-may-2025, ROS should work on linux only. Checking a few other source, it seems that some community efforts support OSX as well, so having both platforms seems reasonable. With this, I think having it for the `cc_ros_{binary, library}` macros makes sense. Of course, you could leave it to the user to specify this every time, but I cannot imagine any situation where this would work on other platforms.

Taking a closer look to the deps, having it in the macros only would allow us to remove it from `roscpp` and `roscpp_traits` since both depend on at least one `cc_ros_library`.

Reply:
If building of a (cc_ros_*) target fails, my opinion is that user/dev should be informed what is incompatible. Top-level macro targets such as cc_ros_* are not good enough IMO -- too high level. It's identified that `roscpp_lib` should be constrained (IMO) and that one should pop up in the error message.

Reply:
Ok, so you would say that we leave `roscpp_core` and `roscpp_traits` with the constraints, but remove them from `cc_ros_*`? Note that this means, I would have to add them to all `cc_ros_*` usages in `rules_ros` instead (we already tried that). There are 7 usages of `cc_ros_library` and 6 usages of `cc_ros_binary`. Do we want this?

Reply:
> Note that this means, I would have to add them to all cc_ros_* usages in rules_ros instead (we already tried that). There are 7 usages of cc_ros_library and 6 usages of cc_ros_binary. Do we want this?

AFAIK, target_compatible_with is propagated by Bazel and this would be unnecessary.

Reply:
Correct, it is propagated through. That's why I suggest adding it only to the ros_cc_* macros and removing it from roscpp_core and traits (since they get it from the macros). This way, we would only have the constraint at 2 places. What do you think about that?

Reply:
In my eyes, ros_cc_* macros are not "problematic" targets, but roscpp_lib and (maybe) roscpp_traits. Annotating those two with target_compatible_with is enough. If someone wants to compile for an incompatible target, they should see that roscpp_lib is the problematic one.

Reply:
I tried removing it from the `ros_cc_*` macros. Now I fail at https://github.com/mvukov/rules_ros/blob/358a0ae782804c3aa86e2f762723b5015f306b5d/ros/repositories/ros_comm.BUILD.bazel#L109. Can we convert this error reporting into a target compatible with as well?

To move on with this PR, I am removing it for the  `ros_cc_*` macros

Reply:
Fine with me to remove `no_match_error`, but then you'll need to provide a default substitution (same as for macos)?

Please also use `macos` instead of `osx` in target_compatible_with, that's a more recent one.

Reply:
Perhaps we don't need target_compatible_with in config_h (expand_template), but just a default substitution.

Reply:
@mvukov I replaced osx with macos to make this PR ready. I would suggest we move the discussion about removin `no_match_error` to a follow up PR